app/api/ada/hardware-wallet/connectTrezorAdaWallet.js

```diff
@@ -0,0 +1,232 @@
+import { createAdaHardwareWallet } from '../adaWallet';
+import { createHardwareWalletAccount } from '../adaAccount';
+import type { ConfigType } from '../../../../config/config-types';
+
+type AddressInfo = { address: string, isUsed: boolean, index: number };
+
+declare var CONFIG: ConfigType;
+const addressScanSize = CONFIG.app.addressScanSize;
+const addressRequestSize = CONFIG.app.addressRequestSize;
+
+/** Creates a new/empty ada wallet
+ * Fetches all related addresses
+ * Caches the fetched address results locally
+ * @returns a new AdaWallet
+ */
+export async function connectTrezorAdaWallet({
+  walletInitData
+}: AdaHardwareWalletParams): Promise<AdaWallet> {
+  try {
+    Logger.debug('connectTrezorAdaWallet::connectTrezorAdaWallet called');
+
+    // create ada wallet object for hardware wallet
+    const [adaWallet] = createAdaHardwareWallet({ walletInitData });
+    // create crypto account object for hardware wallet
+    // eslint-disable-next-line max-len
+    const cryptoAccount = createHardwareWalletAccount(walletInitData.cwHardwareInfo.publicMasterKey);
+
+    // fetch all addresses
+    // Note: addresses may be skipped if generated by a wallet that doesn't follow bip44
+
+    // fetch all External addresses
+    const externalAddressesToSave =
+      await _discoverAllAddressesFrom(cryptoAccount, 'External', -1, addressScanSize, addressRequestSize);
+
+    // fetch all Internal addresses
+    const internalAddressesToSave =
+      await _discoverAllAddressesFrom(cryptoAccount, 'Internal', -1, addressScanSize, addressRequestSize);
+
+    // store wallet related addresses to lovefieldDatabase
+    if (externalAddressesToSave.length !== 0 || internalAddressesToSave.length !== 0) {
+      await Promise.all([
+        saveAsAdaAddresses(cryptoAccount, externalAddressesToSave, 'External'),
+        saveAsAdaAddresses(cryptoAccount, internalAddressesToSave, 'Internal')
+      ]);
+    } else {
+      // no related addresses found, give it a new address and save it to lovefieldDatabase
+      await newAdaAddress(cryptoAccount, [], 'External');
+    }
+
+    // save crypto account to the local storage
+    saveCryptoAccount(cryptoAccount);
+
+    // save ada wallet to the local storage
+    saveAdaWallet(adaWallet);
+
+    // It's a success, we are done
+    // 1. creating wallet and crypto account
+    // 2. storing wallet and crypto account to the local storage
+    // 3. storing wallet related addresses to lovefieldDatabase
+    Logger.debug('connectTrezorAdaWallet::connectTrezorAdaWallet success');
+    return adaWallet;
+  } catch (error) {
+    Logger.error(`connectTrezorAdaWallet::connectTrezorAdaWallet error: ${stringifyError(error)}`);
+    throw error;
+  }
+}
+
+// FIXME : repeated in all wallet creation, try to make it reusable
+
+/** Repeatedly scan addresses until there is a contiguous block of `scanSize` addresses unused
+ * @returns all scanned addresses
+ */
+async function _discoverAllAddressesFrom(
```

Comment:
this needs to be fixed before merging. We can’t have duplicated code :-/

Reply:
https://github.com/Emurgo/yoroi-frontend/commit/31f8da8a36cd23e11be1ce94dbf5e7a6cbcb9929